core/integration/remotecache_test.go

```diff
@@ -3,252 +3,193 @@ package core
 import (
 	"context"
 	"fmt"
-	"os"
-	"os/exec"
-	"sync"
+	"strings"
 	"testing"
-	"time"
 
 	"dagger.io/dagger"
 	"github.com/moby/buildkit/identity"
 	"github.com/stretchr/testify/require"
+	"github.com/tidwall/gjson"
 )
 
-func TestRemoteCacheRegistry(t *testing.T) {
-	ctx, cancel := context.WithCancel(context.Background())
-	defer cancel()
-
-	registryContainerName := runRegistryInDocker(ctx, t)
-	getClient := func() *dagger.Client {
-		c, stop := runSeparateEngine(ctx, t, nil, map[string]string{
-			"_EXPERIMENTAL_DAGGER_CACHE_CONFIG": "type=registry,ref=127.0.0.1:5000/test-cache,mode=max",
-		}, "container:"+registryContainerName)
-		t.Cleanup(func() {
-			stop()
-		})
-		return c
+func getDevEngine(ctx context.Context, c *dagger.Client, cache *dagger.Container, cacheName, cacheEnv string, index uint8) (devEngine *dagger.Container, endpoint string, err error) {
+	id := identity.NewID()
+	networkCIDR := fmt.Sprintf("10.%d.0.0/16", 100+index)
+	// This loads the engine.tar file from the host into the container, that was set up by
+	// internal/mage/engine.go:test. This is used to spin up additional dev engines.
+	devEngineTar := c.Host().Directory("/dagger-dev/", dagger.HostDirectoryOpts{Include: []string{"engine.tar"}}).File("engine.tar")
+
+	devEngine = c.Container().Import(devEngineTar)
+	entrypoint, err := devEngine.File("/usr/local/bin/dagger-entrypoint.sh").Contents(ctx)
+	if err != nil {
+		return nil, "", err
 	}
+	entrypoint = strings.ReplaceAll(entrypoint, "10.88.0.0/16", networkCIDR)
+	entrypoint = strings.ReplaceAll(entrypoint, "dagger-dev", fmt.Sprintf("remote-cache-%d", index))
 
-	pipelineOutput := func(c *dagger.Client) string {
-		output, err := c.Container().From("alpine:3.17").WithExec([]string{
-			"sh", "-c", "head -c 128 /dev/random | sha256sum",
-		}).Stdout(ctx)
-		require.NoError(t, err)
-		return output
-	}
+	devEngine = devEngine.WithNewFile("/usr/local/bin/dagger-entrypoint.sh", dagger.ContainerWithNewFileOpts{
+		Contents: entrypoint,
+	})
 
-	/*
-		1. Start a registry for storing the cache
-		2. Start two independent engines from empty cache that are configured to use the registry as remote cache backend
-		3. Run an exec w/ output from /dev/random in the first engine
-		4. Close the first engine's client, flushing the remote cache for the session
-		5. Run the same exec in the second engine, verify it imports the cache and output the same value as the first engine
-	*/
-	clientA := getClient()
-	clientB := getClient()
-	outputA := pipelineOutput(clientA)
-	require.NoError(t, clientA.Close())
-	outputB := pipelineOutput(clientB)
-	require.Equal(t, outputA, outputB)
-}
+	devEngine = devEngine.
+		WithServiceBinding(cacheName, cache).
+		WithExposedPort(1234, dagger.ContainerWithExposedPortOpts{Protocol: dagger.Tcp}).
+		WithEnvVariable("_EXPERIMENTAL_DAGGER_CACHE_CONFIG", cacheEnv).
+		WithEnvVariable("ENGINE_ID", id).
+		WithMountedCache("/var/lib/dagger", c.CacheVolume("dagger-dev-engine-state-"+id)).
```

Comment:
I just realized that once we are running this on a persistent engine then we need to make sure this cache mount is never reused across runs. Randomizing the name should be a super simple way
```suggestion
		WithMountedCache("/var/lib/dagger", c.CacheVolume("dagger-dev-engine-state-"+identity.NewID())).
```